### PR TITLE
Don't rebuild LLVM for BOLT optimization

### PR DESCRIPTION
**Pull request by nikic**
Currently, we perform a separate rustc/LLVM build with BOLT instrumentation. However BOLT implementation works on compiled artifacts, so I don't think there is any reason to do a full rebuild. Instead, we should perform the full build, and then at the end instrument libLLVM.so, profile, and optimize libLLVM.so, without performing any rustc/LLVM rebuilds.

cc @Kobzol @Mark-Simulacrum

### CODE REVIEW

**Comment by Kobzol** [comment]
I think that it's a good idea to move BOLT optimization outside of bootstrap (or, at least, outside of the LLVM build step) to avoid messing with bootstrap caches, so I like this direction.

I would suggest a slightly different workflow though, and perform the intermediate build steps manually, instead of running `dist` in the middle of the pipeline:
1) Stage 3: build PGO optimized LLVM with relocations
2) Stash away a copy of the built LLVM library and then BOLT instrument it
3) Gather profiles
4) (do not delete LLVM directory, just keep it)
5) BOLT optimize the stashed uninstrumented copy and replace the instrumented LLVM library with it
6) Run `dist`, which shouldn't need to rebuild LLVM again

I hope that this would work, as it's still a bit unclear to me how is the built LLVM exactly used (is rustc only linked to it? or is it used to actually build LLVM? or does this differ for individual stages of rustc build - 0/1/2?)

We could optimize this further by building a PGO optimized LLVM sooner, in stage 2, and then stash it away, to avoid one rustc rebuild.

**Comment by nikic** [comment]
Tried to implement something along those lines but can't test locally right now, so...

@bors try

**Comment by nikic** [comment]
Let's verify whether we actually end up using the optimized LLVM...

@rust-timer build 3b5a8f7fa72f048331b445d463f6758214ee06cd

From the build log, it looks like we do save one LLVM build (3 instead of 4), but we do still appear to rebuild rustc in the dist stage.

New build stats:
```
2023-02-01T00:13:46.2607661Z ---------------------------------------------------------
2023-02-01T00:13:46.2608080Z Build rustc (LLVM PGO generate):        1733.73s (21.89%)
2023-02-01T00:13:46.2608511Z Gather profiles (LLVM PGO):              604.11s ( 7.63%)
2023-02-01T00:13:46.2608897Z Build rustc (rustc PGO generate):        688.46s ( 8.69%)
2023-02-01T00:13:46.2609276Z Gather profiles (rustc PGO):            1056.09s (13.34%)
2023-02-01T00:13:46.2609697Z Build rustc (rustc PGO use, LLVM PGO use):      1533.21s (19.36%)
2023-02-01T00:13:46.2610092Z Bolt instrument LLVM:                    258.50s ( 3.26%)
2023-02-01T00:13:46.2610532Z Gather profiles (LLVM BOLT):             717.22s ( 9.06%)
2023-02-01T00:13:46.2610912Z Bolt optimize LLVM:                       37.05s ( 0.47%)
2023-02-01T00:13:46.2611264Z Dist rustc:                             1290.43s (16.30%)
2023-02-01T00:13:46.2611601Z Total duration:                         7918.80s
2023-02-01T00:13:46.2612196Z ---------------------------------------------------------
```
Build stats from a previous merge (https://pipelines.actions.githubusercontent.com/serviceHosts/1c7a4aeb-bdca-499c-8aff-37881d0d775b/_apis/pipelines/1/runs/26914/signedlogcontent/31?urlExpires=2023-02-01T08%3A40%3A53.6122741Z&urlSigningMethod=HMACV1&urlSignature=YZMD8t53%2FZHCOazFEnkffUUqJSDVxYM5HjufP3dvF6g%3D):
```

2023-02-01T07:46:52.1086106Z ---------------------------------------------------------
2023-02-01T07:46:52.1086531Z Build rustc (LLVM PGO):                 1460.78s (14.52%)
2023-02-01T07:46:52.1086928Z Gather profiles (LLVM PGO):              515.45s ( 5.12%)
2023-02-01T07:46:52.1087340Z Build rustc (rustc PGO):                 740.59s ( 7.36%)
2023-02-01T07:46:52.1087741Z Gather profiles (rustc PGO):            1131.28s (11.24%)
2023-02-01T07:46:52.1088130Z Build rustc (LLVM BOLT):                2189.50s (21.76%)
2023-02-01T07:46:52.1088534Z Gather profiles (LLVM BOLT):             891.15s ( 8.86%)
2023-02-01T07:46:52.1088910Z Final build:                            3134.70s (31.15%)
2023-02-01T07:46:52.1089562Z Total duration:                        10063.45s
2023-02-01T07:46:52.1090202Z ---------------------------------------------------------
```

So this does look worthwhile, but it's not the best we can do.

**Comment by nikic** [comment]
Okay, looks like the optimized LLVM did not get picked up :(

**Comment by Kobzol** [comment]
I also had some problems with locating the right LLVM file. I think that the file being used for `dist` might not actually be the one that you use, but a (hard?)link to this file in some other directory. The hard link is broken, since `llvm-bolt` creates a new file at the original path, and that might be the reason why it doesn't work.

We should print the output of `get_built_llvm_lib_path()` in the previous bootstrap BOLT implementation to see what is the correct LLVM lib path to optimize.

**Comment by nikic** [comment]
Running `./x.py build -vv` locally I see:
```
Copy "/home/npopov/repos/rust/build/x86_64-unknown-linux-gnu/llvm/build/lib/libLLVM-15-rust-dev.so" to "/home/npopov/repos/rust/build/x86_64-unknown-linux-gnu/stage1/lib/libLLVM-15-rust-dev.so"
```
So it looks like the libLLVM is taken from `llvm/build/lib`, while here I'm overwriting the one in `llvm/lib` here.

**Comment by nikic** [comment]
Let's see if this works...

@bors try @rust-timer queue

**Comment by Kobzol** [comment]
One disadvantage of moving BOLT outside of bootstrap is that the paths to these files will be a bit fragile in the script  (another one is that it will be more complicated to perform a BOLT build manually/locally, but that's not a big deal). If we change it in bootstrap, it will also need to be changed in the Python script. But I suppose that it should be visible pretty quickly, because perf. should regress if it fails to optimize some file.

File: src/bootstrap/dist.rs [hunk 1]
```diff
@@ -2220,10 +2220,6 @@ impl Step for ReproducibleArtifacts {
             tarball.add_file(path, ".", 0o644);
             added_anything = true;
         }
-        if let Some(path) = builder.config.llvm_bolt_profile_use.as_ref() {
```
Comment (Kobzol) [inline src/bootstrap/dist.rs:2223]: Another disadvantage of putting BOLT into Python is that we lose reproducibility of the profiles here :( I wonder if we could add a bootstrap step on top of LLVM that would create the BOLTed libraries in another place (in order not to interfere with the regular LLVM build step, and to sidestep its cache), and thus still keep BOLT as a first class citizen, with proper access to LLVM library paths, and with storage of profiles, but at the same time while avoiding the extra LLVM (re)build.

**Comment by Kobzol** [comment]
Great, now it looks that it has worked! I still wonder if it's a good idea to move BOLT completely outside of bootstrap, because some of the mentioned problem (brittle LLVM path, no straightforward access to BOLT profiles, difficult to use BOLT locally). I think that having it as a  different build step in bootstrap might be better.

But I wouldn't block this PR on it, it speeds up CI.

**Comment by nikic** [comment]
@Kobzol I think the ideal here might be to have a callback step in bootstrap which allow post-processing the artifacts. We could then hook into there to optimize the final stage2 artifacts, which avoids some dependence on details like what gets copied where, and would also make sure that there is no additional rustc rebuild. This would probably also be convenient for optimizing rustc itself with BOLT, as otherwise it might be tricky to convince the build system to actually use the optimized binaries.

This could either be a generic callback (run an extra command after assembling the stage2 compiler), or it could be something BOLT-specific, where we would perform BOLT instrumentation and optimization in bootstrap, and only make the callback provide the merged profile data file. Not sure which of those would be better, really.

Anyway, I'm on PTO from tomorrow, so I'd only be able to experiment with something like that once I get back.

**Comment by nikic** [comment]
Closing this in favor of #107723.

**Pull request closed without merging**